ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -271,4 +489,376 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+                  processNewPut(newScanResults, originalPut);
+                }
+              } else {
+                Offset previousIndexSegmentOffset = index.getIndexSegments().size() > 1 ? index.getIndexSegments()
+                    .lowerKey(indexSegment.getStartOffset()) : index.getStartOffset();
+                FileSpan searchSpan = new FileSpan(index.getStartOffset(), previousIndexSegmentOffset);
+                originalPut = index.findKey(entry.getKey(), searchSpan);
+              }
+              if (originalPut != null && !originalPut.isFlagSet(IndexValue.Flags.Delete_Index)) {
+                // put exists and there are no multiple deletes
+                processNewDelete(newScanResults, indexValue, originalPut);
+              }
+            } else {
+              // put record
+              processNewPut(newScanResults, indexValue);
+            }
+          }
+        }
+      }
+    }
+
+    /**
+     * Process a {@link List} of {@link IndexEntry} belonging to the same {@link IndexSegment} to populate container and
+     * log segment buckets. The function is called in reverse chronological order. That is, newest {@link IndexSegment}
+     * to older ones.
+     * @param indexSegment the {@link IndexSegment} where the index entries belong to
+     * @param indexEntries a {@link List} of {@link IndexEntry} to be processed
+     * @param deletedKeys a {@link Map} of processed deleted keys to their corresponding deletion time in ms
+     * @throws StoreException
+     */
+    private void processIndexSegmentEntriesBackward(IndexSegment indexSegment, List<IndexEntry> indexEntries,
+        Map<StoreKey, Long> deletedKeys) throws StoreException {
+      // valid index entries wrt container reference time
+      List<IndexEntry> validIndexEntries =
+          getValidIndexEntries(indexSegment, indexEntries, newScanResults.containerForecastStartTimeMs, deletedKeys);
+      processEntriesForContainerBucket(newScanResults, validIndexEntries, deletedKeys);
+      // valid index entries wrt log segment reference time
+      validIndexEntries =
+          getValidIndexEntries(indexSegment, indexEntries, newScanResults.logSegmentForecastStartTimeMs, deletedKeys);
+      processEntriesForLogSegmentBucket(newScanResults, validIndexEntries, deletedKeys);
+    }
+
+    /**
+     * Get a {@link List} of filtered {@link IndexEntry} from a given {@link IndexSegment} whose elements all have a
+     * {@link Offset} that is strictly less than the given endOffset.
+     * @param indexSegment the {@link IndexSegment} to get the index entries from
+     * @param endOffset the {@link Offset} that defines the filter boundary
+     * @return a {@link List} of {@link IndexEntry} whose elements all have a {@link Offset} that is strictly less than
+     * the given endOffset
+     * @throws StoreException
+     */
+    private List<IndexEntry> getFilteredIndexEntries(IndexSegment indexSegment, Offset endOffset)
```

Comment:
name it as "getIndexEntriesUntillOffset" or "getIndexEntriesLessThanOffset" or "getIndexEntriesUptoOffset". Current name gives a notion that we are filtering only for entries having the passed in offset sorta.